### PR TITLE
add mt5 to ORTConfigManager conf list

### LINKED ISSUE #321: MT5 Support
Big Thanks for the awesome Optimum library, 
MT5 is a Massive Transformers And Supports Many Languages, but it has a slow inference time
for speedup MT5 inference time, We need some post-training methods, and Optimum will help more!

Please Add MT5 Support for Optimum! Thanks

### PR DESCRIPTION
**Pull request by ChainYo**
# What does this PR do?

Add MT5 to `ORTConfigManager`.

Fixes #321 

I re-arranged in alphabetical order all available models. I can put it back like it was if needed. 🤗

@JingyaHuang 

*Aside from this PR, I was wondering if opening an issue like https://github.com/huggingface/transformers/issues/16308 for implementing all available onnx models in the ORTConfigManager could be nice?*

### CODE REVIEW

**Comment by JingyaHuang** [comment]
Hi @ChainYo, thank you for adding mt5 in transformers and in optimum!

Sorting the models in alphabetical order makes sense to me. Can you also add a test of mt5 on `tests/onnxruntime/test_optimization.py` with the model `hf-internal-testing/tiny-random-mt5`? Thanks!

**Comment by ChainYo** [comment]
Same as above, I re-arranged in alphabetical order all tested architectures.

Because `MT5` has no sequence-classification feature, I introduced a new argument: `model_feature`. Let me know if it's ok or not.
But it could help in the future with other models not designed for `sequence-classification`.

File: tests/onnxruntime/test_optimization.py [hunk 1]
```diff
@@ -55,26 +56,27 @@ def test_save_and_load(self):
...
+        (ElectraForSequenceClassification, "hf-internal-testing/tiny-random-electra", "sequence-classification"),
+        (GPT2ForSequenceClassification, "hf-internal-testing/tiny-random-gpt2", "sequence-classification"),
+        (MT5ForConditionalGeneration, "hf-internal-testing/tiny-random-mt5", "seq2seq-lm"),
+        (RobertaForSequenceClassification, "hf-internal-testing/tiny-random-roberta", "sequence-classification"),
+        (XLMRobertaForSequenceClassification, "hf-internal-testing/tiny-xlm-roberta", "sequence-classification"),
     )
 
     @parameterized.expand(SUPPORTED_ARCHITECTURES_WITH_MODEL_ID)
-    def test_optimize(self, model_cls, model_name):
+    def test_optimize(self, model_cls, model_name, model_feature):
```
Comment (ChainYo) [inline tests/onnxruntime/test_optimization.py:71]: Here is the new `str` argument for the model feature.

File: tests/onnxruntime/test_optimization.py [hunk 2]
```diff
@@ -55,26 +56,27 @@ def test_save_and_load(self):
...
-        (BigBirdForSequenceClassification, "hf-internal-testing/tiny-random-big_bird"),
+        (BartForSequenceClassification, "hf-internal-testing/tiny-random-bart", "sequence-classification"),
+        (BertForSequenceClassification, "hf-internal-testing/tiny-random-bert", "sequence-classification"),
+        (BigBirdForSequenceClassification, "hf-internal-testing/tiny-random-big_bird", "sequence-classification"),
+        (DistilBertForSequenceClassification, "hf-internal-testing/tiny-random-distilbert", "sequence-classification"),
+        (ElectraForSequenceClassification, "hf-internal-testing/tiny-random-electra", "sequence-classification"),
+        (GPT2ForSequenceClassification, "hf-internal-testing/tiny-random-gpt2", "sequence-classification"),
+        (MT5ForConditionalGeneration, "hf-internal-testing/tiny-random-mt5", "seq2seq-lm"),
+        (RobertaForSequenceClassification, "hf-internal-testing/tiny-random-roberta", "sequence-classification"),
+        (XLMRobertaForSequenceClassification, "hf-internal-testing/tiny-xlm-roberta", "sequence-classification"),
```
Comment (ChainYo) [inline tests/onnxruntime/test_optimization.py:67]: I added the `sequence-classification` feature for others and `seq2seq-lm` for the MT5 model.

**Review by JingyaHuang — COMMENTED** [review]
The refactoring looks neat to me, pinging @regisss and @michaelbenayoun for some inputs.

**Review by regisss — APPROVED** [review]
LGTM, thanks for this @ChainYo!!

**Comment by regisss** [comment]
The build PR doc test should pass now

**Comment by regisss** [comment]
As for opening an issue for implementing all available ONNX models in the ORTConfigManager, why not. IT could be a good way to track what models still have to be added. What do you think @JingyaHuang @echarlaix @mfuntowicz @fxmarty ?

**Comment by JingyaHuang** [comment]
Besides, it seems that the CI of the optimization has one failed test:
```
2022-08-10T14:39:30.5123260Z =========================== short test summary info ============================
2022-08-10T14:39:30.5123701Z FAILED onnxruntime/test_optimization.py::ORTOptimizerTest::test_optimize_6 - ...
```
I can't see details about where it failed. Can you run the test locally to get the complete log of the failure? Thank you!

**Comment by JingyaHuang** [comment]
> As for opening an issue for implementing all available ONNX models in the ORTConfigManager, why not. It could be a good way to track which models still have to be added. What do you think @JingyaHuang @echarlaix @mfuntowicz @fxmarty ?

IMO, it makes sense to keep them all in the test as far as it won't slow down the CI. When it turns to be too slow, we can archive a part of them as slow tests.

**Comment by regisss** [comment]
> > As for opening an issue for implementing all available ONNX models in the ORTConfigManager, why not. It could be a good way to track which models still have to be added. What do you think @JingyaHuang @echarlaix @mfuntowicz @fxmarty ?
> 
> IMO, it makes sense to keep them all in the test as far as it won't slow down the CI. When it turns to be too slow, we can archive a part of them as slow tests.

I was actually talking about the suggestion of opening an issue such as [this one](https://github.com/huggingface/transformers/issues/16308) to keep track of the models to add to `ORTConfigManager`.

**Comment by JingyaHuang** [comment]
@regisss oh sorry, I misunderstood it. I think that is a good idea! Both users and contributors could have a better idea about what has been integrated, what need to be integrated, and probably how to do that.

**Comment by regisss** [comment]
> @regisss oh sorry, I misunderstood it. I think that is a good idea! Both users and contributors could have a better idea about what has been integrated, what needs to be integrated, and probably some references on how to do that.

Yes exactly :)

@ChainYo Feel free to create such an issue :+1:

**Comment by ChainYo** [comment]
> > @regisss, oh sorry, I misunderstood it. I think that is a good idea! Both users and contributors could have a better idea about what has been integrated, what needs to be integrated, and probably some references on how to do that.
> 
> Yes exactly :)
> 
> @ChainYo Feel free to create such an issue +1

Cool, I will open it right now!

> Besides, it seems that the CI of the optimization has one failed test:
> 
> ```
> 2022-08-10T14:39:30.5123260Z =========================== short test summary info ============================
> 2022-08-10T14:39:30.5123701Z FAILED onnxruntime/test_optimization.py::ORTOptimizerTest::test_optimize_6 - ...
> ```
> 
> I can't see details about where it failed. Can you run the test locally to get the complete log of the failure? Thank you!

I will take a look at the failure message.

**Comment by JingyaHuang** [comment]
Hi @ChainYo,

Thanks for explaining, that's clear. Indeed, as the CI leverages the latest transformers, it won't pass until your contribution of MT5 ONNX config makes its way to a release of transformers.

**Comment by customer101** [comment]
Hi @ChainYo,
I really appreciate the work done to support mT5 on optimum, this saves me a lot of time.
I have faced some issues after running this code :
```
model_path = "./mt5_model"
tokenizer = T5Tokenizer.from_pretrained(model_path)
model = ORTModelForSeq2SeqLM.from_pretrained(model_path, from_transformers=True)
```
it fails at `optimum/onnxruntime/modeling_seq2seq.py` :
```
        # Export the decoder with the past key values
        if use_cache:
            export(
                preprocessor=tokenizer,
                model=decoder_with_lm_head,
                config=onnx_config_decoder_with_past,
                opset=onnx_opset,
                output=save_dir.joinpath(ONNX_DECODER_WITH_PAST_NAME),
            )
```
Error:
```
RuntimeError: Sizes of tensors must match except in dimension 2. Expected size 85 but got size 64 for tensor number 1 in the list.
```
If I set `use_cache=false`, it works fine. Does setting it to false will degrade the speed-up expected from onnx optimizations?

**Comment by ChainYo** [comment]
> Hi @ChainYo, I really appreciate the work done to support mT5 on optimum, this saves me a lot of time. I have faced some issues after running this code :
> 
> ```
> model_path = "./mt5_model"
> tokenizer = T5Tokenizer.from_pretrained(model_path)
> model = ORTModelForSeq2SeqLM.from_pretrained(model_path, from_transformers=True)
> ```
> 
> it fails at `optimum/onnxruntime/modeling_seq2seq.py` :
> 
> ```
>         # Export the decoder with the past key values
>         if use_cache:
>             export(
>                 preprocessor=tokenizer,
>                 model=decoder_with_lm_head,
>                 config=onnx_config_decoder_with_past,
>                 opset=onnx_opset,
>                 output=save_dir.joinpath(ONNX_DECODER_WITH_PAST_NAME),
>             )
> ```
> 
> Error:
> 
> ```
> RuntimeError: Sizes of tensors must match except in dimension 2. Expected size 85 but got size 64 for tensor number 1 in the list.
> ```
> 
> If I set `use_cache=false`, it works fine. Does setting it to false will degrade the speed-up expected from onnx optimizations?



> Hi @ChainYo, I really appreciate the work done to support mT5 on optimum, this saves me a lot of time. I have faced some issues after running this code :
> 
> ```
> model_path = "./mt5_model"
> tokenizer = T5Tokenizer.from_pretrained(model_path)
> model = ORTModelForSeq2SeqLM.from_pretrained(model_path, from_transformers=True)
> ```
> 
> it fails at `optimum/onnxruntime/modeling_seq2seq.py` :
> 
> ```
>         # Export the decoder with the past key values
>         if use_cache:
>             export(
>                 preprocessor=tokenizer,
>                 model=decoder_with_lm_head,
>                 config=onnx_config_decoder_with_past,
>                 opset=onnx_opset,
>                 output=save_dir.joinpath(ONNX_DECODER_WITH_PAST_NAME),
>             )
> ```
> 
> Error:
> 
> ```
> RuntimeError: Sizes of tensors must match except in dimension 2. Expected size 85 but got size 64 for tensor number 1 in the list.
> ```
> 
> If I set `use_cache=false`, it works fine. Does setting it to false will degrade the speed-up expected from onnx optimizations?

HI @customer101, could you check your model's `config.json` file and tell me the architecture used? Did you convert the MT5 model to ONNX with the main branch of Transformers?

Are you using it in a pipeline or with vanilla PyTorch? Where do you add the `use_cache=False` argument?

**Comment by customer101** [comment]
I used transformers main branch, the model is vanilla PyTorch,
this line:
```
model = ORTModelForSeq2SeqLM.from_pretrained(model_path, from_transformers=True, use_cache=False)
```
does the conversion, then when I save the model it dumps 2 onnx models(encoder and decoder).
Here is the `config.json`:
```
{
  "_name_or_path": "mt5_model",
  "architectures": [
    "MT5ForConditionalGeneration"
  ],
  "d_ff": 1024,
  "d_kv": 64,
  "d_model": 512,
  "decoder_start_token_id": 0,
  "dropout_rate": 0.1,
  "early_stopping": true,
  "eos_token_id": 1,
  "feed_forward_proj": "gated-gelu",
  "initializer_factor": 1.0,
  "is_encoder_decoder": true,
  "layer_norm_epsilon": 1e-06,
  "max_length": 256,
  "model_type": "mt5",
  "num_beams": 3,
  "num_decoder_layers": 8,
  "num_heads": 6,
  "num_layers": 8,
  "pad_token_id": 0,
  "relative_attention_num_buckets": 32,
  "tie_word_embeddings": false,
  "tokenizer_class": "T5Tokenizer",
  "torch_dtype": "float32",
  "transformers_version": "4.17.0",
  "use_cache": true,
  "vocab_size": 250112
}
```
I noticed that mt5 optimizations are not implemented yet in onnxruntime, so I think I won't gain much speed-up by converting to onnx.
Let me know if you need further information, good luck @ChainYo .

**Comment by ChainYo** [comment]
HI @JingyaHuang @regisss 

Transformers v4.22.0 has included the MT5 OnnxConfig and stuff. Can we relaunch tests and see if all is fine here?

**Comment by JingyaHuang** [comment]
Hi @ChainYo,

Sure, before launching the tests again, can you do a rebase with the main branch of Optimum? Thank you!

**Comment by ChainYo** [comment]
> Sure, before launching the tests again, can you do a rebase with the main branch of Optimum? Thank you!

It's done @JingyaHuang

Optimizer behavior has changed, so adding a feature argument is not needed anymore!

**Comment by ChainYo** [comment]
I fixed a typo for tests.

I don't know if `ORTModelForCustomTasks` is the best one, but I didn't find any `ORTModelForConditionalGeneration`. Or maybe I should import the MT5 specific one?

**Comment by JingyaHuang** [comment]
Hi @ChainYo, the current `ORTModelForCustomTasks` doesn't support the sequence-to-sequence model yet, it shall be `ORTModelForSeq2SeqLM` in this case.

As @customer101 observes, for the moment, there is a problem with the dimension of past-key-value, so you need to set `use_cache=False` for the time being as a workaround. 

Here is the snippet:
```
from optimum.onnxruntime import ORTModelForSeq2SeqLM
model = ORTModelForSeq2SeqLM.from_pretrained("google/mt5-small", from_transformers=True, use_cache=False)
```

**Comment by JingyaHuang** [comment]
Hi @customer101 ,

Thanks for reporting the issue with `use_past`. It seems that there is an issue while exporting decoder_with_past.

`past_key_values` contains pre-computed keys and values for attention which helps to speed up sequential decoding.  Can you open an separate issue concerning the bug, so that we can discuss and tackle it together?

**Comment by ChainYo** [comment]
Hi @JingyaHuang,

I see there is two problems according to the error tests in the CI.
- First: about `evaluate`: 

```
ModuleNotFoundError: No module named 'evaluate'
```

- Second: about `mt5` KeyError

```

**Review by JingyaHuang — CHANGES_REQUESTED** [review]
Hi @ChainYo, 

Thanks for tracking the PR.

To pass the CIs, here are several modifications to apply:

* For the benchmark suite test, can you rebase your branch? The evaluate module has been added since #384 
* For the MT5 support, the name of hidden size shall be `d_model`.
* ONNX Runtime optimization supports bert-like, gpt-like and bart-like models. As t5/mt5 are encoder-decoder models, I suggest setting the model type as "bart" to benefit from [some extra fusions](https://github.com/microsoft/onnxruntime/blob/189aef2beaecdbdd3a68fbf569d9891c11d9777c/onnxruntime/python/tools/transformers/onnx_model_bart.py#L307-L308).
* And we recently reduced the size of some models to accelerate the CIs, can you use "hf-internal-testing/tiny-random-onnx-mt5" for the test?

With the modifications beyond, the PR should be good to go!

**Review by JingyaHuang — APPROVED** [review]

**Review by JingyaHuang — APPROVED** [review]
Thanks for iterating the PR, LGTM!

**Pull request merged**